torch/distributed/optim/functional_adadelta.py

```diff
@@ -22,6 +22,7 @@ def __init__(
         rho: float = 0.9,
         eps: float = 1e-6,
         weight_decay: float = 0.0,
+        allow_empty_param_list: bool = False,
```

Comment:
curious, where and how is this used? If there is no params, does not work if we just don't create a functional optimizer instance?

Reply:
@rohan-varma's use case for `allow_empty_param_list` is slightly different than mine. For me, it is simpler in the `ZeroRedundancyOptimizer` implementation if we do not have to have `if ... else ...` in places to check if `self.optim` is `None` (i.e. uninitialized) or not. This happens for example when overlapping DDP with ZeRO, there are fewer DDP buckets than ranks, and the approach uses 1 ZeRO bucket for 1 DDP bucket. Passing this `allow_empty_param_list` argument is effectively for code convenience.

For @rohan-varma, it seems from https://github.com/pytorch/pytorch/pull/61678 that his implementation prefers to initialize the functional optimizer as empty and populate its set of parameters on the fly in `step_param()` as the parameters are seen. His use case is probably a stronger motivation for including this change; it just happens that `ZeroRedundancyOptimizer` may also benefit. (He already added it for `_FunctionalSGD`; this is just adding it for the other functional optimizers.)

Reply:
Yeah, the main reason is for functional optimizer overlap with DDP, we create the optimizer upfront (i.e. when communication hook is registered), but are not actually aware and don't want to hold model parameters just yet. We only get parameters to pass in when the comm. hook is invoked with a bucket, so at that point we'd still like to call `step_param` with the given parameter and its gradient without having saved the parameters earlier.

Reply:
So comm hook can only be registered if we already initialize optimizer (instead of directly install the hooks into class)? If that's the case, I guess we should introduce this parameter. @andwgu Could we make this parameter to be private with `_` prepended? It's not clear to me what's the idea solution of this yet but I think it's always good to make it private and if we stay on this we can switch to public at a later time.